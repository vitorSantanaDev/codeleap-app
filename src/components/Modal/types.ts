import { ReactNode } from 'react'

export type ModalProps = {
  children: ReactNode
  isVisible: boolean
}
