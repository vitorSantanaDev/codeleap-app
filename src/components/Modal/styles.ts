import styled, { css } from 'styled-components'
import { ModalProps } from './types'
import media from 'styled-media-query'

export type WrapperProps = Pick<ModalProps, 'isVisible'>

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, isVisible }) => css`
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    height: 100vh;
    overflow: hidden;
    position: fixed;

    z-index: ${theme.layers.modal};
    opacity: ${isVisible ? '1' : '0'};
    transition: opacity 0.3s ease-in-out;
    pointer-events: ${isVisible ? 'all' : 'none'};
    background-color: ${theme.colors.overlay_color};

    display: flex;
    align-items: center;
    justify-content: center;

    ${media.lessThan('medium')`
      padding: 0px ${theme.spacings.small};
    `}
  `}
`

export const ContentWrapper = styled.div`
  ${({ theme }) => css`
    width: 100%;
    max-width: 66rem;
    padding: ${theme.spacings.small};
    border-radius: ${theme.border.radius};
    background-color: ${theme.colors.white};
    border: 1px solid ${theme.colors.gray};
  `}
`
