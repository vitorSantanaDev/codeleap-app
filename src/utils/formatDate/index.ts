import {
  format,
  parseISO,
  minutesToHours,
  secondsToMinutes,
  differenceInSeconds
} from 'date-fns'

export const formattingStringWithTheTimePassedBetweenTwoDate = (
  hours: number,
  minutes: number,
  seconds: number
): string => {
  if (hours < 1 && minutes < 1) {
    return `${seconds} seconds ago`
  }

  let formattedTime = ''

  if (hours > 0) {
    formattedTime = `${hours} ${hours > 1 ? 'hours' : 'hour'} ago`
  } else if (hours < 1) {
    formattedTime = `${minutes} minutes ago`
  }

  return formattedTime
}

export const rangeOfTimePassedBetweenTwoDates = (
  start: string,
  end: string
): string => {
  const startDate = parseISO(start)
  const endDate = parseISO(end)

  const allSeconds = differenceInSeconds(endDate, startDate)
  const allMinutes = secondsToMinutes(allSeconds)

  const minutesInSeconds = allMinutes * 60
  const seconds = allSeconds - minutesInSeconds

  const hours = minutesToHours(allMinutes)

  const hoursInMinutes = hours * 60

  const minutes = allMinutes - hoursInMinutes

  return hours > 24
    ? format(new Date(startDate), 'dd/MM/yyyy')
    : formattingStringWithTheTimePassedBetweenTwoDate(hours, minutes, seconds)
}
