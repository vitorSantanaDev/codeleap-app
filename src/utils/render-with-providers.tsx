import { Provider } from 'react-redux'
import React, { PropsWithChildren } from 'react'

import theme from 'styles/theme'
import { ThemeProvider } from 'styled-components'
import { RenderOptions, render } from '@testing-library/react'

import { RootState } from 'redux/store'
import { userReducer } from 'redux/actions/user-slice'
import { PreloadedState, configureStore } from '@reduxjs/toolkit'
import { QueryClient, QueryClientProvider } from 'react-query'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  store?: any
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = configureStore({
      reducer: { user: userReducer },
      preloadedState: initialState
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <QueryClientProvider client={new QueryClient()}>
        <Provider store={store}>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </Provider>
      </QueryClientProvider>
    )
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
