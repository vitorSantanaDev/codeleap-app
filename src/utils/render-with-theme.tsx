import React, { ReactElement } from 'react'

import theme from 'styles/theme'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

export const renderWithTheme = (component: ReactElement) =>
  render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
