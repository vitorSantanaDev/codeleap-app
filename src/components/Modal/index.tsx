import React from 'react'

import { ModalProps } from './types'

import * as S from './styles'

const Modal: React.FC<ModalProps> = ({ children, isVisible }) => {
  return (
    <S.Wrapper
      aria-label="modal"
      role="presentation"
      isVisible={isVisible}
      aria-hidden={!isVisible}
    >
      <S.ContentWrapper>{children}</S.ContentWrapper>
    </S.Wrapper>
  )
}

export default Modal
