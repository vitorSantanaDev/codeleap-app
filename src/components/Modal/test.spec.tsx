import React from 'react'

import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/render-with-theme'

import Modal from '.'

describe('<Modal />', () => {
  it('should render modal content', () => {
    renderWithTheme(
      <Modal isVisible>
        <h1>Modal</h1>
      </Modal>
    )

    const modalWrapper = screen.getByRole('presentation', { name: /modal/i })
    expect(modalWrapper).toBeInTheDocument()

    expect(modalWrapper.getAttribute('aria-hidden')).toBe('false')
    expect(modalWrapper).toHaveStyle({ opacity: 1 })

    const heading = screen.getByRole('heading', { name: /modal/i })
    expect(heading).toBeInTheDocument()
  })

  it('should not render modal when isVisible is false', () => {
    renderWithTheme(
      <Modal isVisible={false}>
        <h1>Modal</h1>
      </Modal>
    )

    const modalWrapper = screen.getByRole('presentation', { hidden: true })

    expect(modalWrapper.getAttribute('aria-hidden')).toBe('true')
    expect(modalWrapper).toHaveStyle({ opacity: 0 })

    const heading = screen.queryByRole('heading', { name: /modal/i })
    expect(heading).not.toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const { container } = renderWithTheme(
      <Modal isVisible={true}>
        <h1>Modal</h1>
      </Modal>
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
