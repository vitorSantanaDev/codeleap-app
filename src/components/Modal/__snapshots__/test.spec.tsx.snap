// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should match snapshot 1`] = `
.c0 {
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  height: 100vh;
  overflow: hidden;
  position: fixed;
  z-index: 40;
  opacity: 1;
  -webkit-transition: opacity 0.3s ease-in-out;
  transition: opacity 0.3s ease-in-out;
  pointer-events: all;
  background-color: rgba(119,119,119,0.8);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  width: 100%;
  max-width: 66rem;
  padding: 2.4rem;
  border-radius: 0.8rem;
  background-color: #FFFFFF;
  border: 1px solid #999999;
}

@media (max-width:768px) {
  .c0 {
    padding: 0px 2.4rem;
  }
}

<div
  aria-hidden="false"
  aria-label="modal"
  class="c0"
  role="presentation"
>
  <div
    class="c1"
  >
    <h1>
      Modal
    </h1>
  </div>
</div>
`;
